### PR TITLE
[#11463] Add other teams/students in section as option in feedback paths

### LINKED ISSUE #11463: Allow 'other teams/students in section' as an option in feedback paths
Current:
image

A user suggested that we add the options:
`other students in the same section`
`other teams in the same section`

Reason: if students are supposed to evaluate others in their own section only, it can be quite inconvenient to choose the correct students when the class is large e.g., the user who suggested this has 800+ students and 170 teams in the course.

I think the feature is useful to have but not sure how much effort the implementation will take.

### PR DESCRIPTION
**Pull request by tjtanjin**
Fixes #11463  

**Outline of Solution**

- Add dropdown option for other teams/students in section in feedback paths.

Example views are provided below:

This is how the 2 new options look like:

creator_perspective

For the remaining views, they are based of Francis who is a student within this demo course - notice the section "Tutorial Group 3" and team "Team A" which Francis is in:

all_students

In the feedback session where question 2 is set to be only for students of the same section, this is the view Francis sees when giving feedback:

charlie_perspective_giver_1

In the feedback session where question 2 is set to be only for students of the same section, this is the response view Francis sees:

charlie_perspective_recipient_1

In the feedback session where question 3 is set to be only for teams of the same section, this is the view Francis sees when giving feedback:

charlie_perspective_giver_2

In the feedback session where question 3 is set to be only for teams of the same section, this is the response view Francis sees:

charlie_perspective_recipient_2

### CODE REVIEW

**Comment by damithc** [comment]
@tjtanjin I'm OK with the UI. 
I'll leave others to check the logic. But remember to make it work when the respondent have to pick `up to n recipients` from the same section. e.g., give feedback up to 5 students in the same section.

**Comment by tjtanjin** [comment]
> @tjtanjin I'm OK with the UI. I'll leave others to check the logic. But remember to make it work when the respondent have to pick `up to n recipients` from the same section. e.g., give feedback up to 5 students in the same section.

Hi Prof Damith, may I check on how the selection up `up to n recipients` should work? I initially tested this on my work branch and noticed issues using this option. Following which, I checked out to the master branch to test this option but it seems to also be showing unexpected behaviour as well. I am not sure if I may have modified anything locally but this is what I've observed on the master branch:

Question 1 is being used for testing here, I tried changing the maximum number of students but even though the value supposedly saves, a refresh shows otherwise.

https://user-images.githubusercontent.com/43908963/141615322-94e251f6-2e69-4a7b-beb3-289ec2a1d533.mp4

Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.

https://user-images.githubusercontent.com/43908963/141615403-7be5cd80-477b-4751-9153-b6166bdd433f.mp4

Strangely, I cannot spot any noticeable difference between the 2 question types. In fact, question 1 was initially having the exact same issue as question 2 and simply suddenly changed its behaviour as I continued testing. I would like to double check if this is an issue I am facing locally and if so, could anyone possibly point me towards the right direction to debugging this? I am under the impression that if I am within the master branch (with ./gradlew generateTypes and npm run build executed), this should be working as intended. Thank you!

**Comment by damithc** [comment]
> Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.

I couldn't reproduce this on the production server. I've created an account for you in the production server. You can try to test is over there too.

**Comment by tjtanjin** [comment]
> > Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.
> 
> I couldn't reproduce this on the production server. I've created an account for you in the production server. You can try to test is over there too.

Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

**Comment by damithc** [comment]
> Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

Yes, it could be a bug introduced more recently.

**Comment by wkurniawan07** [comment]
Between latest `release` and latest `master` there is only one PR merged (#11423), however based on the nature of the bug reported, it makes a lot of sense if the cause is that PR. @xinyee20 can you spend some time to look into this?

**Comment by damithc** [comment]
> Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

@tjtanjin Please report this as a separate bug. As it is already in the master and breaks a core function, we need to fix it soon.

**Comment by madanalogy** [comment]
@tjtanjin the question edit bug should have been resolved with https://github.com/TEAMMATES/teammates/pull/11480, will leave this PR as ongoing as it seems that you are still working on addressing some comments. Feel free to switch the tag if I was mistaken or once the PR is ready for review otherwise.

**Comment by tjtanjin** [comment]
> @tjtanjin the question edit bug should have been resolved with #11480, will leave this PR as ongoing as it seems that you are still working on addressing some comments. Feel free to switch the tag if I was mistaken or once the PR is ready for review otherwise.

I have merged the master branch and fixed the custom number of recipients. Tested it out and it is working as expected so the PR is ready for review! :)

**Comment by xinyee20** [comment]
> Between latest `release` and latest `master` there is only one PR merged (#11423), however based on the nature of the bug reported, it makes a lot of sense if the cause is that PR. @xinyee20 can you spend some time to look into this?

Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.

**Comment by madanalogy** [comment]
> Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.

Should have been resolved with https://github.com/TEAMMATES/teammates/pull/11480

**Comment by xinyee20** [comment]
> > Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.
> 
> Should have been resolved with #11480

@wkurniawan07 thank you for fixing it!

**Review by daongochieu2810 — APPROVED** [review]

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 1]
```diff
@@ -455,6 +503,26 @@ public boolean sessionHasQuestions(String feedbackSessionName, String courseId)
...
+                        Const.InstructorPermissions.CAN_SUBMIT_SESSION_IN_SECTIONS)) {
+                    // instructor can only see teams in allowed sections for him/her
+                    continue;
+                }
+                // Ensure student('s team) does not evaluate own team.
+                if (!giverTeam.equals(team.getKey())) {
+                    // recipientEmail doubles as team name in this case.
+                    recipients.put(team.getKey(), team.getKey());
+                }
+            }
```
Comment (jianhandev) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java:524]: This part is identical from the previous case, might have room for refactoring. Perhaps let the `TEAMS` case fall through to  the `TEAMS_IN_SAME_SECTION` case and check for `generateOptionsFor == FeedbackParticipantType.TEAMS_IN_SAME_SECTION` / `generateOptionsFor == FeedbackParticipantType.TEAMS`. Similarly for the `STUDENTS_IN_SAME_SECTION` case above.

**Review by jianhandev — CHANGES_REQUESTED** [review]
Overall looks good, in some places the use of fallthrough may help to reduce duplication of code.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java (hunk 1 above)
Comment (tjtanjin) [reply src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: I have updated the code as you have suggested, thank you!

**Review by jianhandev — APPROVED** [review]

**Pull request merged**